<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Predicting Single-cell Enhancer Gene Interactions by Integrating Prior Hi-C Information • SCEGHiC</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Predicting Single-cell Enhancer Gene Interactions by Integrating Prior Hi-C Information">
<meta name="description" content="SCEG-HiC predicts interactions between genes and enhancers using multi-ome (paired scRNA-seq and scATAC-seq or only scATAC-seq) sequencing data and three-dimensional omics data (the bulk average Hi-C). The approach uses weighted graphical lasso (wlasso) model, taking the average Hi-C data as prior knowledge, to predict cell specific interactions between genes and enhancers based on the correlation between genes and enhancers.">
<meta property="og:description" content="SCEG-HiC predicts interactions between genes and enhancers using multi-ome (paired scRNA-seq and scATAC-seq or only scATAC-seq) sequencing data and three-dimensional omics data (the bulk average Hi-C). The approach uses weighted graphical lasso (wlasso) model, taking the average Hi-C data as prior knowledge, to predict cell specific interactions between genes and enhancers based on the correlation between genes and enhancers.">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-11"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-11');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SCEGHiC</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="articles/pbmc_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregrate</a></li>
    <li><a class="dropdown-item" href="articles/pbmc_multiomic_CD4T.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</a></li>
    <li><a class="dropdown-item" href="articles/mouse_skin_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of mouse skin-aggregate</a></li>
    <li><a class="dropdown-item" href="articles/human_covid_19_scATAC_seq_mono.html">SCEG-HiC on only scATAC-seq data of human COVID 19-Mono</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC/discussions">Discussions</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="sceg-hicsingle-cell-enhancer-gene-interactions-by-integrating-priori-hi-c-information">SCEG-HiC(single-cell enhancer gene interactions by integrating priori Hi-C information)<a class="anchor" aria-label="anchor" href="#sceg-hicsingle-cell-enhancer-gene-interactions-by-integrating-priori-hi-c-information"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end -->
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>SCEG-HiC predicts interactions between genes and enhancers using multi-ome (paired scRNA-seq and scATAC-seq or only scATAC-seq) sequencing data and three-dimensional omics data (the bulk average Hi-C). The approach uses weighted graphical lasso (wglasso) model, taking the average Hi-C data as priori knowledge, to predict cell specific interactions between genes and enhancers based on the correlation between genes and enhancers.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="required-software">Required software<a class="anchor" aria-label="anchor" href="#required-software"></a>
</h3>
<p>SCEG-HiC runs in the <a href="https://www.r-project.org/" class="external-link">R statistical computing environment</a>. You will need R version 4.1.0 or higher, <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> version 3.14 or higher, and Seurat 4.0 or higher to have access to the latest features.</p>
<p>To install Bioconductor, open R and run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Next, install a few Bioconductor dependencies that aren’t automatically installed:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'BiocGenerics'</span>, <span class="st">'DelayedArray'</span>, <span class="st">'DelayedMatrixStats'</span>,</span>
<span>                       <span class="st">'limma'</span>, <span class="st">'lme4'</span>, <span class="st">'S4Vectors'</span>, <span class="st">'SingleCellExperiment'</span>,</span>
<span>                       <span class="st">'SummarizedExperiment'</span>, <span class="st">'batchelor'</span>, <span class="st">'HDF5Array'</span>,</span>
<span>                       <span class="st">'terra'</span>, <span class="st">'ggrastr'</span>,<span class="st">'Gviz'</span>, <span class="st">'rtracklayer'</span>,<span class="st">'GenomeInfoDb'</span>,<span class="st">'GenomicRanges'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Installation of other dependencies</p>
<ul>
<li>Install Signac pacakge : <code>devtools::install_github("timoast/signac", ref = "develop")</code>. Please check <a href="https://stuartlab.org/signac/" class="external-link">here</a> if you encounter any issue.</li>
<li>Install Cicero package: <code>devtools::install_github("cole-trapnell-lab/cicero-release", ref = "monocle3")</code>. Please check <a href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3/#installing-cicero" class="external-link">here</a> if you encounter any issue.</li>
</ul>
<p>Now, you can install the development version of SCEG-HiC from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"wuwei77lx/SCEGHiC"</span><span class="op">)</span></span></code></pre></div>
<p>If you wish to install CRAN of SCEG-HiC, execute:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install released version from CRAN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"SCEGHiC"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="testing-the-installation">Testing the installation<a class="anchor" aria-label="anchor" href="#testing-the-installation"></a>
</h3>
<p>To ensure that SCEG-HiC was installed correctly, start a new R session and run:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wuwei77lx/SCEGHiC" class="external-link">SCEGHiC</a></span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="the-bulk-average-hi-c">The bulk average Hi-C<a class="anchor" aria-label="anchor" href="#the-bulk-average-hi-c"></a>
</h2>
<p>The celltypes used by human for averaging are: GM12878, NHEK, HMEC, RPE1, THP1, IMR90, HUVEC, HCT116, K562, KBM7.</p>
<p>The celltypes used by mouse for averaging are: mESC1, mESC2, CH12LX, CH12F3, fiber, epithelium, B.</p>
<p>Generate the bulk average Hi-C using the <a href="https://www.nature.com/articles/s41588-019-0538-0" class="external-link">Activity by Contact (ABC)</a> model’s <code>makeAverageHiC.py</code>.</p>
<p>Average Hi-C data can be downloaded from:</p>
<ul>
<li>The human avg HiC file here: <a href="https://www.encodeproject.org/files/ENCFF134PUN/@@download/ENCFF134PUN.bed.gz" class="external-link uri">https://www.encodeproject.org/files/ENCFF134PUN/@@download/ENCFF134PUN.bed.gz</a>
</li>
</ul>
<p>Extract the human bulk average Hi-C using the <a href="https://www.nature.com/articles/s41588-019-0538-0" class="external-link">Activity by Contact (ABC)</a> model’s <code>extract_avg_hic.py</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">python</span> workflow/scripts/extract_avg_hic.py <span class="at">--avg_hic_bed_file</span> ../ENCFF134PUN.bed.gz <span class="at">--output_dir</span> ../</span></code></pre></div>
<ul>
<li>The mouse avg HiC file here: <a href="https://zenodo.org/record/14849886" class="external-link">10.5281/zenodo.14849886</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="quickstart">Quickstart<a class="anchor" aria-label="anchor" href="#quickstart"></a>
</h2>
<p>This is a basic example which shows you how to solve a common problem:</p>
<p>Aggregated with paired scRNA-seq and scATAC-seq data</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wuwei77lx/SCEGHiC" class="external-link">SCEGHiC</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="co">## basic example code</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">multiomic_small</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Preprocess data</span></span>
<span><span class="va">SCEGdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">multiomic_small</span>, k_neigh <span class="op">=</span> <span class="fl">5</span>, max_overlap <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating aggregated data</span></span>
<span><span class="co">#&gt; Aggregating cluster 0</span></span>
<span><span class="co">#&gt; Sample cells randomly.</span></span>
<span><span class="co">#&gt; have 11 samples</span></span>
<span><span class="co">#&gt; Aggregating cluster 1</span></span>
<span><span class="co">#&gt; Sample cells randomly.</span></span>
<span><span class="co">#&gt; have 11 samples</span></span>
<span></span>
<span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRABD2A"</span>, <span class="st">"GNLY"</span>, <span class="st">"MFSD6"</span>, <span class="st">"CTLA4"</span>, <span class="st">"LCLAT1"</span>, <span class="st">"NCK2"</span>, <span class="st">"GALM"</span>, <span class="st">"TMSB10"</span>, <span class="st">"ID2"</span>, <span class="st">"CXCR4"</span><span class="op">)</span></span>
<span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCEGHiC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate weight</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/calculateHiCWeights.html">calculateHiCWeights</a></span><span class="op">(</span><span class="va">SCEGdata</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="va">gene</span>, averHicPath <span class="op">=</span> <span class="va">fpath</span><span class="op">)</span></span>
<span><span class="co">#&gt; Successfully obtained 10 TSS loci of genes from chromosome 2.</span></span>
<span><span class="co">#&gt; Calculate weight of  TRABD2A</span></span>
<span><span class="co">#&gt; Calculate weight of  GNLY</span></span>
<span><span class="co">#&gt; Calculate weight of  MFSD6</span></span>
<span><span class="co">#&gt; Calculate weight of  CXCR4</span></span>
<span><span class="co">#&gt; Calculate weight of  CTLA4</span></span>
<span><span class="co">#&gt; Calculate weight of  LCLAT1</span></span>
<span><span class="co">#&gt; Calculate weight of  NCK2</span></span>
<span><span class="co">#&gt; Calculate weight of  ID2</span></span>
<span><span class="co">#&gt; Calculate weight of  GALM</span></span>
<span><span class="co">#&gt; Calculate weight of  TMSB10</span></span>
<span></span>
<span><span class="co"># Run model</span></span>
<span><span class="va">results_SCEGHiC</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/Run_SCEG_HiC.html">Run_SCEG_HiC</a></span><span class="op">(</span><span class="va">SCEGdata</span>, <span class="va">weight</span>, focus_gene <span class="op">=</span> <span class="va">gene</span><span class="op">)</span></span>
<span><span class="co">#&gt; The predicted genes are 10 in total.</span></span>
<span><span class="co">#&gt; Run model of  TRABD2A</span></span>
<span><span class="co">#&gt; Run model of  GNLY</span></span>
<span><span class="co">#&gt; Run model of  MFSD6</span></span>
<span><span class="co">#&gt; Run model of  CXCR4</span></span>
<span><span class="co">#&gt; Run model of  CTLA4</span></span>
<span><span class="co">#&gt; Run model of  LCLAT1</span></span>
<span><span class="co">#&gt; Run model of  NCK2</span></span>
<span><span class="co">#&gt; Run model of  ID2</span></span>
<span><span class="co">#&gt; Run model of  GALM</span></span>
<span><span class="co">#&gt; Run model of  TMSB10</span></span>
<span></span>
<span><span class="co"># Visualize the links</span></span>
<span><span class="fu"><a href="reference/connections_Plot.html">connections_Plot</a></span><span class="op">(</span><span class="va">results_SCEGHiC</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="st">"CTLA4"</span>, cutoff <span class="op">=</span> <span class="fl">0.01</span>, gene_anno <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-example1-1.png" width="100%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Coverage plot and visualize the links of CTLA4</span></span>
<span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"multiomic_small_atac_fragments.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"SCEGHiC"</span><span class="op">)</span></span>
<span><span class="va">frags</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CreateFragmentObject.html" class="external-link">CreateFragmentObject</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">fpath</span>, cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">multiomic_small</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing hash</span></span>
<span><span class="fu"><a href="https://stuartlab.org/signac/reference/Fragments.html" class="external-link">Fragments</a></span><span class="op">(</span><span class="va">multiomic_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">frags</span></span>
<span><span class="fu"><a href="reference/coverPlot.html">coverPlot</a></span><span class="op">(</span><span class="va">multiomic_small</span>, focus_gene <span class="op">=</span> <span class="st">"CTLA4"</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, assay <span class="op">=</span> <span class="st">"peaks"</span>, SCEG_HiC_Result <span class="op">=</span> <span class="va">results_SCEGHiC</span>, SCEG_HiC_cutoff <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">#&gt;   suppressWarnings() to suppress this warning.)</span></span></code></pre></div>
<p><img src="reference/figures/README-example1-2.png" width="100%"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-conda-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: CentOS Linux 7 (Core)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS/LAPACK: /home/liangxuan/conda/envs/test/lib/libopenblasp-r0.3.28.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggplot2_3.5.1                     BSgenome.Hsapiens.UCSC.hg38_1.4.5</span></span>
<span><span class="co">#&gt;  [3] BSgenome_1.74.0                   rtracklayer_1.66.0               </span></span>
<span><span class="co">#&gt;  [5] BiocIO_1.16.0                     Biostrings_2.74.1                </span></span>
<span><span class="co">#&gt;  [7] XVector_0.46.0                    EnsDb.Hsapiens.v86_2.99.0        </span></span>
<span><span class="co">#&gt;  [9] ensembldb_2.30.0                  AnnotationFilter_1.30.0          </span></span>
<span><span class="co">#&gt; [11] GenomicFeatures_1.58.0            AnnotationDbi_1.68.0             </span></span>
<span><span class="co">#&gt; [13] Biobase_2.66.0                    GenomicRanges_1.58.0             </span></span>
<span><span class="co">#&gt; [15] GenomeInfoDb_1.42.1               IRanges_2.40.1                   </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.44.0                  BiocGenerics_0.52.0              </span></span>
<span><span class="co">#&gt; [19] Seurat_5.2.0                      SeuratObject_5.0.2               </span></span>
<span><span class="co">#&gt; [21] sp_2.1-4                          Signac_1.14.9001                 </span></span>
<span><span class="co">#&gt; [23] SCEGHiC_0.0.0.9000               </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] R.methodsS3_1.8.2           dichromat_2.0-0.1          </span></span>
<span><span class="co">#&gt;   [3] progress_1.2.3              urlchecker_1.0.1           </span></span>
<span><span class="co">#&gt;   [5] nnet_7.3-20                 goftest_1.2-3              </span></span>
<span><span class="co">#&gt;   [7] vctrs_0.6.5                 spatstat.random_3.3-2      </span></span>
<span><span class="co">#&gt;   [9] digest_0.6.37               png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [11] ggrepel_0.9.6               styler_1.10.3              </span></span>
<span><span class="co">#&gt;  [13] deldir_2.0-4                parallelly_1.41.0          </span></span>
<span><span class="co">#&gt;  [15] MASS_7.3-64                 reshape2_1.4.4             </span></span>
<span><span class="co">#&gt;  [17] httpuv_1.6.15               withr_3.0.2                </span></span>
<span><span class="co">#&gt;  [19] xfun_0.50                   ellipsis_0.3.2             </span></span>
<span><span class="co">#&gt;  [21] survival_3.8-3              memoise_2.0.1              </span></span>
<span><span class="co">#&gt;  [23] commonmark_1.9.2            rcmdcheck_1.4.0            </span></span>
<span><span class="co">#&gt;  [25] profvis_0.4.0               zoo_1.8-12                 </span></span>
<span><span class="co">#&gt;  [27] pbapply_1.7-2               R.oo_1.27.0                </span></span>
<span><span class="co">#&gt;  [29] Formula_1.2-5               prettyunits_1.2.0          </span></span>
<span><span class="co">#&gt;  [31] KEGGREST_1.46.0             promises_1.3.2             </span></span>
<span><span class="co">#&gt;  [33] httr_1.4.7                  restfulr_0.0.15            </span></span>
<span><span class="co">#&gt;  [35] globals_0.16.3              fitdistrplus_1.2-2         </span></span>
<span><span class="co">#&gt;  [37] ps_1.8.1                    rstudioapi_0.17.1          </span></span>
<span><span class="co">#&gt;  [39] UCSC.utils_1.2.0            miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt;  [41] generics_0.1.3              base64enc_0.1-3            </span></span>
<span><span class="co">#&gt;  [43] processx_3.8.5              curl_6.0.1                 </span></span>
<span><span class="co">#&gt;  [45] zlibbioc_1.52.0             polyclip_1.10-7            </span></span>
<span><span class="co">#&gt;  [47] GenomeInfoDbData_1.2.13     SparseArray_1.6.0          </span></span>
<span><span class="co">#&gt;  [49] xopen_1.0.1                 xtable_1.8-4               </span></span>
<span><span class="co">#&gt;  [51] stringr_1.5.1               desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [53] evaluate_1.0.3              S4Arrays_1.6.0             </span></span>
<span><span class="co">#&gt;  [55] BiocFileCache_2.14.0        hms_1.1.3                  </span></span>
<span><span class="co">#&gt;  [57] irlba_2.3.5.1               colorspace_2.1-1           </span></span>
<span><span class="co">#&gt;  [59] filelock_1.0.3              ROCR_1.0-11                </span></span>
<span><span class="co">#&gt;  [61] reticulate_1.40.0           spatstat.data_3.1-4        </span></span>
<span><span class="co">#&gt;  [63] magrittr_2.0.3              lmtest_0.9-40              </span></span>
<span><span class="co">#&gt;  [65] later_1.4.1                 lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [67] spatstat.geom_3.3-4         future.apply_1.11.3        </span></span>
<span><span class="co">#&gt;  [69] scattermore_1.2             XML_3.99-0.17              </span></span>
<span><span class="co">#&gt;  [71] cowplot_1.1.3               matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt;  [73] RcppAnnoy_0.0.22            Hmisc_5.2-2                </span></span>
<span><span class="co">#&gt;  [75] pillar_1.10.1               nlme_3.1-166               </span></span>
<span><span class="co">#&gt;  [77] cicero_1.3.9                compiler_4.4.2             </span></span>
<span><span class="co">#&gt;  [79] RSpectra_0.16-2             stringi_1.8.4              </span></span>
<span><span class="co">#&gt;  [81] tensor_1.5                  minqa_1.2.8                </span></span>
<span><span class="co">#&gt;  [83] SummarizedExperiment_1.36.0 devtools_2.4.5             </span></span>
<span><span class="co">#&gt;  [85] GenomicAlignments_1.42.0    plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [87] crayon_1.5.3                abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [89] bit_4.5.0.1                 dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;  [91] fastmatch_1.1-6             NCmisc_1.2.0               </span></span>
<span><span class="co">#&gt;  [93] codetools_0.2-20            monocle3_1.3.7             </span></span>
<span><span class="co">#&gt;  [95] bslib_0.8.0                 slam_0.1-55                </span></span>
<span><span class="co">#&gt;  [97] biovizBase_1.54.0           plotly_4.10.4              </span></span>
<span><span class="co">#&gt;  [99] mime_0.12                   splines_4.4.2              </span></span>
<span><span class="co">#&gt; [101] Rcpp_1.0.14                 fastDummies_1.7.4          </span></span>
<span><span class="co">#&gt; [103] dbplyr_2.5.0                interp_1.1-6               </span></span>
<span><span class="co">#&gt; [105] knitr_1.49                  blob_1.2.4                 </span></span>
<span><span class="co">#&gt; [107] lme4_1.1-36                 fs_1.6.5                   </span></span>
<span><span class="co">#&gt; [109] listenv_0.9.1               checkmate_2.3.2            </span></span>
<span><span class="co">#&gt; [111] Rdpack_2.6.2                pkgbuild_1.4.5             </span></span>
<span><span class="co">#&gt; [113] Gviz_1.50.0                 tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [115] Matrix_1.6-5                callr_3.7.6                </span></span>
<span><span class="co">#&gt; [117] tweenr_2.0.3                pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [119] tools_4.4.2                 cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [121] R.cache_0.16.0              rbibutils_2.3              </span></span>
<span><span class="co">#&gt; [123] RSQLite_2.3.9               viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [125] DBI_1.2.3                   fastmap_1.2.0              </span></span>
<span><span class="co">#&gt; [127] rmarkdown_2.29              scales_1.3.0               </span></span>
<span><span class="co">#&gt; [129] grid_4.4.2                  usethis_3.1.0              </span></span>
<span><span class="co">#&gt; [131] ica_1.0-3                   Rsamtools_2.22.0           </span></span>
<span><span class="co">#&gt; [133] sass_0.4.9                  patchwork_1.3.0            </span></span>
<span><span class="co">#&gt; [135] FNN_1.1.4.1                 dotCall64_1.2              </span></span>
<span><span class="co">#&gt; [137] VariantAnnotation_1.52.0    RANN_2.6.2                 </span></span>
<span><span class="co">#&gt; [139] rpart_4.1.24                farver_2.1.2               </span></span>
<span><span class="co">#&gt; [141] reformulas_0.4.0            yaml_2.3.10                </span></span>
<span><span class="co">#&gt; [143] VGAM_1.1-12                 roxygen2_7.3.2             </span></span>
<span><span class="co">#&gt; [145] latticeExtra_0.6-30         MatrixGenerics_1.18.1      </span></span>
<span><span class="co">#&gt; [147] foreign_0.8-88              cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [149] purrr_1.0.2                 lifecycle_1.0.4            </span></span>
<span><span class="co">#&gt; [151] uwot_0.2.2                  sessioninfo_1.2.2          </span></span>
<span><span class="co">#&gt; [153] backports_1.5.0             BiocParallel_1.40.0        </span></span>
<span><span class="co">#&gt; [155] gtable_0.3.6                rjson_0.2.23               </span></span>
<span><span class="co">#&gt; [157] ggridges_0.5.6              progressr_0.15.1           </span></span>
<span><span class="co">#&gt; [159] parallel_4.4.2              testthat_3.2.3             </span></span>
<span><span class="co">#&gt; [161] jsonlite_1.8.9              RcppHNSW_0.6.0             </span></span>
<span><span class="co">#&gt; [163] bitops_1.0-9                bit64_4.5.2                </span></span>
<span><span class="co">#&gt; [165] assertthat_0.2.1            brio_1.1.5                 </span></span>
<span><span class="co">#&gt; [167] Rtsne_0.17                  glasso_1.11                </span></span>
<span><span class="co">#&gt; [169] spatstat.utils_3.1-2        jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [171] spatstat.univar_3.1-1       R.utils_2.12.3             </span></span>
<span><span class="co">#&gt; [173] lazyeval_0.2.2              shiny_1.10.0               </span></span>
<span><span class="co">#&gt; [175] htmltools_0.5.8.1           sctransform_0.4.1          </span></span>
<span><span class="co">#&gt; [177] rappdirs_0.3.3              glue_1.8.0                 </span></span>
<span><span class="co">#&gt; [179] spam_2.11-0                 httr2_1.0.7                </span></span>
<span><span class="co">#&gt; [181] RCurl_1.98-1.16             rprojroot_2.0.4            </span></span>
<span><span class="co">#&gt; [183] jpeg_0.1-10                 gridExtra_2.3              </span></span>
<span><span class="co">#&gt; [185] boot_1.3-31                 igraph_2.0.3               </span></span>
<span><span class="co">#&gt; [187] R6_2.5.1                    tidyr_1.3.1                </span></span>
<span><span class="co">#&gt; [189] SingleCellExperiment_1.28.1 RcppRoll_0.3.1             </span></span>
<span><span class="co">#&gt; [191] labeling_0.4.3              cluster_2.1.8              </span></span>
<span><span class="co">#&gt; [193] pkgload_1.4.0               nloptr_2.1.1               </span></span>
<span><span class="co">#&gt; [195] DelayedArray_0.32.0         tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt; [197] ProtGenerics_1.38.0         htmlTable_2.4.3            </span></span>
<span><span class="co">#&gt; [199] ggforce_0.4.2               xml2_1.3.6                 </span></span>
<span><span class="co">#&gt; [201] future_1.34.0               vdiffr_1.0.8               </span></span>
<span><span class="co">#&gt; [203] munsell_0.5.1               KernSmooth_2.23-26         </span></span>
<span><span class="co">#&gt; [205] data.table_1.16.4           htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt; [207] RColorBrewer_1.1-3          biomaRt_2.62.0             </span></span>
<span><span class="co">#&gt; [209] rlang_1.1.4                 spatstat.sparse_3.1-0      </span></span>
<span><span class="co">#&gt; [211] spatstat.explore_3.3-4      remotes_2.5.0              </span></span>
<span><span class="co">#&gt; [213] reader_1.0.6</span></span></code></pre></div>
</details><p>See <a href="https://wuwei77lx.github.io/SCEGHiC/">documentation website</a> for more information!</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>In SCEG-HiC, you can choose to aggregate data or use single-cell type data for analysis.</p>
<ul>
<li><p>aggregate data: average data of similar cells,which are used k-nearest neighbor method.</p></li>
<li><p>single-cell type data: normalize a single-cell type data.</p></li>
</ul>
<p>For more information about how this package has been used with real data, please check the following links:</p>
<ul>
<li><p><a href="https://wuwei77lx.github.io/SCEGHiC/articles/pbmc_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregate</a></p></li>
<li><p><a href="https://wuwei77lx.github.io/SCEGHiC/articles/pbmc_multiomic_CD4T.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</a></p></li>
<li><p><a href="https://wuwei77lx.github.io/SCEGHiC/articles/mouse_skin_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of mouse skin-aggregate</a></p></li>
<li><p><a href="https://wuwei77lx.github.io/SCEGHiC/articles/human_covid_19_scATAC_seq_mono.html">SCEG-HiC on only scATAC-seq data of human COVID 19-Mono</a></p></li>
</ul>
</div>
<div class="section level2">
<h2 id="help">Help<a class="anchor" aria-label="anchor" href="#help"></a>
</h2>
<p>If you have any problems, comments or suggestions, please contact us at XuanLiang (<a href="mailto:liangxuan2022@sinh.ac.cn" class="email">liangxuan2022@sinh.ac.cn</a>).</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/wuwei77lx/SCEGHiC/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/wuwei77lx/SCEGHiC/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SCEGHiC</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Xuan Liang <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xuan Liang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
