---
title: "SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregate"
author: "XuanLiang"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "../../data/human/pbmc")
options(future.globals.maxSize = 20 * 1024^3)
```

In this vignette,we’ll demonstrate  SCEG-HiC’s capability of inferring gene-enhancer by applying it to paired scRNA-seq and scATAC-seq data. In this vignette we'll be using a publicly available 10x Genomic Multiome dataset for human PBMCs,and aggregating the data for processing.

The implementent of SCEG-HiC is seamlessly compatible with the workflow of Seurat/Signac package, SCEG-HiC pipeline consists of the following three major parts:

* Set up a Seurat object.
* Infer gene-enhancer using SCEG-HiC by taking as input  a Seurat object and the bulk average Hi-C.
* Visualize the gene-enhancer.


<details>
  <summary>**View data download code**</summary>

You can download the required data by running the following lines in a shell:

```{sh eval=FALSE}
wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5
wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz
wget https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi
```

</details>

## Load the required libraries
```{r message=FALSE, warning=FALSE}
library(SCEGHiC)
library(Signac)
library(Seurat)
library(EnsDb.Hsapiens.v86)
library(BSgenome.Hsapiens.UCSC.hg38)
library(ggplot2)
library(hdf5r)
```

## Set up a Seurat object

To facilitate easy exploration, ```pbmc_multiomic.rds``` also available from [10.5281/zenodo.14849886](https://zenodo.org/record/14849886).

```{r message=FALSE, warning=FALSE}
# load the RNA and ATAC data
counts <- Read10X_h5("pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5")
fragpath <- "pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz"
```

```{r message=FALSE, warning=FALSE}
# get gene annotations for hg38
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevels(annotation) <- paste0("chr", seqlevels(annotation))

# create a Seurat object containing the RNA data
pbmc <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)

# create ATAC assay and add it to the object
pbmc[["ATAC"]] <- CreateChromatinAssay(
  counts = counts$Peaks,
  sep = c(":", "-"),
  fragments = fragpath,
  annotation = annotation
)
```

```{r}
pbmc
```

```{r fig.width = 18, fig.height = 6,message=FALSE, warning=FALSE}
# Quality control
DefaultAssay(pbmc) <- "ATAC"

pbmc <- NucleosomeSignal(pbmc)
pbmc <- TSSEnrichment(pbmc)

VlnPlot(
  object = pbmc,
  features = c("nCount_RNA", "nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
  ncol = 4,
  pt.size = 0
)
```

```{r}
# filter out low quality cells
pbmc <- subset(
  x = pbmc,
  subset = nCount_ATAC < 100000 &
    nCount_RNA < 25000 &
    nCount_ATAC > 1800 &
    nCount_RNA > 1000 &
    nucleosome_signal < 2 &
    TSS.enrichment > 1
)
pbmc
```

```{r}
# call peaks using MACS2
peaks <- CallPeaks(pbmc)
# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks <- keepStandardChromosomes(peaks, pruning.mode = "coarse")
peaks <- subsetByOverlaps(x = peaks, ranges = blacklist_hg38_unified, invert = TRUE)
# quantify counts in each peak
macs2_counts <- FeatureMatrix(
  fragments = Fragments(pbmc),
  features = peaks,
  cells = colnames(pbmc)
)
# create a new assay using the MACS2 peak set and add it to the Seurat object
pbmc[["peaks"]] <- CreateChromatinAssay(
  counts = macs2_counts,
  fragments = fragpath,
  annotation = annotation
)
pbmc
```

### RNA analysis
```{r results='hide', message=FALSE, warning=FALSE}
DefaultAssay(pbmc) <- "RNA"
pbmc <- SCTransform(pbmc)
pbmc <- RunPCA(pbmc)
```

### ATAC analysis
```{r message=FALSE, warning=FALSE}
DefaultAssay(pbmc) <- "ATAC"
pbmc <- FindTopFeatures(pbmc, min.cutoff = 5)
pbmc <- RunTFIDF(pbmc)
pbmc <- RunSVD(pbmc)
```

### Annotating cell types and Intergration analysis
We'll use an annotated PBMC reference dataset from [Hao et al. (2020)](https://www.biorxiv.org/content/10.1101/2020.10.12.335331v1),
available for download here: https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat.
```{r message=FALSE, warning=FALSE}
library(SeuratDisk)

# load PBMC reference
reference <- LoadH5Seurat("pbmc_multimodal.h5seurat", assays = list("SCT" = "counts"), reductions = "spca")
reference <- UpdateSeuratObject(reference)

DefaultAssay(pbmc) <- "SCT"

# transfer cell type labels from reference to query
transfer_anchors <- FindTransferAnchors(
  reference = reference,
  query = pbmc,
  normalization.method = "SCT",
  reference.reduction = "spca",
  recompute.residuals = FALSE,
  dims = 1:50
)

predictions <- TransferData(
  anchorset = transfer_anchors,
  refdata = reference$celltype.l1,
  weight.reduction = pbmc[["pca"]],
  dims = 1:50
)

pbmc <- AddMetaData(
  object = pbmc,
  metadata = predictions
)

# set the cell identities to the cell type predictions
Idents(pbmc) <- "predicted.id"

levels(pbmc) <- c("CD4 T", "CD8 T", "Mono", "NK", "B", "DC", "other T", "other")
```


Using the weighted nearest neighbor methods in [Seurat v4](https://www.biorxiv.org/content/10.1101/2020.10.12.335331v1),
we can compute a joint neighbor graph that represent both the gene expression and 
DNA accessibility measurements.
```{r fig.width = 6, fig.height = 6,message=FALSE, warning=FALSE}
# build a joint neighbor graph using both assays
pbmc <- FindMultiModalNeighbors(
  object = pbmc,
  reduction.list = list("pca", "lsi"),
  dims.list = list(1:50, 2:40),
  modality.weight.name = "RNA.weight",
  verbose = TRUE
)

# build a joint UMAP visualization
pbmc <- RunUMAP(
  object = pbmc,
  nn.name = "weighted.nn",
  assay = "RNA",
  verbose = TRUE
)

DimPlot(pbmc, label = TRUE, repel = TRUE, reduction = "umap") + NoLegend()
```

```{r include=FALSE}
saveRDS(pbmc, file = "pbmc_multiomic.rds")
```

## Infer gene-enhancer using SCEG-HiC 

### Preprocess data
When preprocessing data with SCEG-HiC, you can choose between clustering or single-cell type analysis. By default, the scATAC-seq counts are binarized.In addition, you can select either paired scRNA-seq and scATAC-seq data or only scATAC-seq data. Here, we have aggregated with paired scRNA-seq and scATAC-seq data.

```{r}
SCEGdata <- process_data(pbmc, max_overlap = 0.5)
```
### Calculate weight
SCEG-HiC takes a long time to calculate weight, and the process becomes slower as more genes are selected. SCEG-HiC provides download websites for the average Hi-C data for both humans and mice. In this case, we select the human average Hi-C data for analysis.

<details>
  <summary>**Select  highly variable features**</summary>
You can select 3000 highly variable features:

```{r eval=FALSE}
gene <- pbmc@assays[["SCT"]]@var.features
weight <- calculateHiCWeights(SCEGdata, species = "Homo sapiens", genome = "hg38", focus_gene = gene, averHicPath = "/picb/bigdata/project/liangxuan/data/human_contact/AvgHiC")
```

</details>

```{r }
# For example, Calculate weight for CD8 T's markers: PDK1
weight <- calculateHiCWeights(SCEGdata, species = "Homo sapiens", genome = "hg38", focus_gene = "PDK1", averHicPath = "/picb/bigdata/project/liangxuan/data/human_contact/AvgHiC")
```

### Run model
SCEG-HiC uses wglasso machine learning method and normalizes the bulk average Hi-C matrix using rank score.
```{r} 
# For example, Run model for CD8 T's markers: PDK1
results_SCEGHiC <- Run_SCEG_HiC(SCEGdata, weight, focus_gene = "PDK1")
```
## Visualize the gene-enhancer 

### Visualize the links

```{r,fig.width = 8, fig.height = 5,message=FALSE,warning=FALSE}
temp <- tempfile()
download.file("https://hgdownload2.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.refGene.gtf.gz", temp)
gene_anno <- rtracklayer::readGFF(temp)
unlink(temp)
# rename some columns to match requirements
gene_anno$chromosome <- gene_anno$seqid
gene_anno$gene <- gene_anno$gene_id
gene_anno$transcript <- gene_anno$transcript_id
gene_anno$symbol <- gene_anno$gene_name
connections_Plot(results_SCEGHiC, species = "Homo sapiens", genome = "hg38", focus_gene = "PDK1", cutoff = 0.01, gene_anno = gene_anno)
```

### Coverage plot and visualize the links of PDK1
Here, the truth of CD8 T is sourced from the [ENCODE database](https://www.encodeproject.org/files/ENCFF063KHP/@@download/ENCFF063KHP.hic), and the correlation is calculated by computing the Pearson correlation coefficient between each gene and enhancer.

```{r, fig.width = 15.5, fig.height = 7,warning=FALSE}
# load the truth of CD8 T data
truth_CD8<-readRDS("truth_CD8.rds")
## load correlation data
correlation<-readRDS("correlation_CD8.rds")
# Coverage plot and visualize the links of PDK1
coverPlot(pbmc, focus_gene = "PDK1", species = "Homo sapiens", genome = "hg38", assay = "peaks", HIC_Result = truth_CD8, SCEG_HiC_Result = results_SCEGHiC, SCEG_HiC_cutoff = 0.01, correlation = correlation, cells = c("CD4 T", "CD8 T"))
```

<details>
  <summary>**Session Info**</summary>

```{r}
sessionInfo()
```

</details>
