<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate promoters, enhancers, and Hi-C matrix of genes — calculateHiCWeights • SCEGHiC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate promoters, enhancers, and Hi-C matrix of genes — calculateHiCWeights"><meta name="description" content="The function obtains enhancers and promoters of each gene, and calculates the bulk average Hi-C matrix of the contact between the gene and enhancers."><meta property="og:description" content="The function obtains enhancers and promoters of each gene, and calculates the bulk average Hi-C matrix of the contact between the gene and enhancers."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-11"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-11');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SCEGHiC</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/pbmc_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregrate</a></li>
    <li><a class="dropdown-item" href="../articles/pbmc_multiomic_CD4T.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</a></li>
    <li><a class="dropdown-item" href="../articles/mouse_skin_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of mouse skin-aggregate</a></li>
    <li><a class="dropdown-item" href="../articles/human_covid_19_scATAC_seq_mono.html">SCEG-HiC on only scATAC-seq data of human COVID 19-Mono</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC/discussions">Discussions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate promoters, enhancers, and Hi-C matrix of genes</h1>
      <small class="dont-index">Source: <a href="https://github.com/wuwei77lx/SCEGHiC/blob/HEAD/R/weight.R" class="external-link"><code>R/weight.R</code></a></small>
      <div class="d-none name"><code>calculateHiCWeights.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function obtains enhancers and promoters of each gene, and calculates the bulk average Hi-C matrix of the contact between the gene and enhancers.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculateHiCWeights</span><span class="op">(</span></span>
<span>  <span class="va">SCEGdata</span>,</span>
<span>  <span class="va">species</span>,</span>
<span>  <span class="va">genome</span>,</span>
<span>  <span class="va">focus_gene</span>,</span>
<span>  <span class="va">averHicPath</span>,</span>
<span>  TSSwindow <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  upstream <span class="op">=</span> <span class="fl">250000</span>,</span>
<span>  downstream <span class="op">=</span> <span class="fl">250000</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-scegdata">SCEGdata<a class="anchor" aria-label="anchor" href="#arg-scegdata"></a></dt>
<dd><p>Preprocessed data for SCEG-HiC.</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>String, indicating the name of the species.It can be either "Homo sapiens"or "Mus musculus".</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>String, indicating the name of the genome assembly.It can be either "hg38",
"hg19", "mm10", or "mm9".</p></dd>


<dt id="arg-focus-gene">focus_gene<a class="anchor" aria-label="anchor" href="#arg-focus-gene"></a></dt>
<dd><p>The focused genes.</p></dd>


<dt id="arg-averhicpath">averHicPath<a class="anchor" aria-label="anchor" href="#arg-averhicpath"></a></dt>
<dd><p>Path to the bulk average Hi-C.</p></dd>


<dt id="arg-tsswindow">TSSwindow<a class="anchor" aria-label="anchor" href="#arg-tsswindow"></a></dt>
<dd><p>Numeric specifying the window size (in base pairs) to pad around either side of each TSS, to fetch gene's promoters.
Default is 1000 bp (so 2000 bp window is drawn around each TSS).</p></dd>


<dt id="arg-upstream">upstream<a class="anchor" aria-label="anchor" href="#arg-upstream"></a></dt>
<dd><p>Numeric specifying the window size (in base pairs) to pad around upstream of each TSS, to fetch gene's enhancers.
Default is 250 kb (so 500 kb window is drawn around each TSS).</p></dd>


<dt id="arg-downstream">downstream<a class="anchor" aria-label="anchor" href="#arg-downstream"></a></dt>
<dd><p>Numeric specifying the window size (in base pairs) to pad around downstream of each TSS, to fetch gene's enhancers.
Default is 250 kb (so 500 kb window is drawn around each TSS).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, should warning and info messages be printed?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with enhancers, promoters, and Hi-C matrices for each <code>focus_gene</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">multiomic_small</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">SCEGdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="process_data.html">process_data</a></span><span class="op">(</span><span class="va">multiomic_small</span>, k_neigh <span class="op">=</span> <span class="fl">5</span>, max_overlap <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating aggregated data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Aggregating cluster 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sample cells randomly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> have 11 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Aggregating cluster 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sample cells randomly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> have 11 samples</span>
<span class="r-in"><span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCEGHiC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRABD2A"</span>, <span class="st">"GNLY"</span>, <span class="st">"MFSD6"</span>, <span class="st">"CTLA4"</span>, <span class="st">"LCLAT1"</span>, <span class="st">"NCK2"</span>, <span class="st">"GALM"</span>, <span class="st">"TMSB10"</span>, <span class="st">"ID2"</span>, <span class="st">"CXCR4"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu">calculateHiCWeights</span><span class="op">(</span><span class="va">SCEGdata</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="va">gene</span>, averHicPath <span class="op">=</span> <span class="va">fpath</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Successfully obtained 10 TSS loci of genes from chromosome 2.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  TRABD2A</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  GNLY</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  MFSD6</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  CXCR4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  CTLA4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  LCLAT1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  NCK2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  ID2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  GALM</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  TMSB10</span>
<span class="r-in"><span><span class="va">weight</span><span class="op">[[</span><span class="st">"CTLA4"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">promoters</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> character(0)</span>
<span class="r-in"><span><span class="va">weight</span><span class="op">[[</span><span class="st">"CTLA4"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">enhancers</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "chr2_203623664_203623982" "chr2_203682042_203682957"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3] "chr2_203730093_203731243" "chr2_203786816_203787404"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5] "chr2_203830381_203830563" "chr2_203932737_203935679"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [7] "chr2_203945543_203945939" "chr2_203992800_203993979"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [9] "chr2_204109915_204110871"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">weight</span><span class="op">[[</span><span class="st">"CTLA4"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">contact</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    element1                 element2        score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;char&gt;                   &lt;char&gt;        &lt;num&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1:    CTLA4 chr2_203623664_203623982 0.0010420455</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2:    CTLA4 chr2_203682042_203682957 0.0009676338</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3:    CTLA4 chr2_203730093_203731243 0.0008481914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4:    CTLA4 chr2_203786816_203787404 0.0014686981</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5:    CTLA4 chr2_203830381_203830563 0.0030630646</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6:    CTLA4 chr2_203932737_203935679 0.0020004300</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xuan Liang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

