<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create SCEG-HiC preprocessed input data — process_data • SCEGHiC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create SCEG-HiC preprocessed input data — process_data"><meta name="description" content="Function to create preprocessed input data for SCEG-HiC.calculateHiCWeights and
Run_SCEG_HiCtakes as input the generated data.This function can generate clustered data
or non clustered data for a specific cell type."><meta property="og:description" content="Function to create preprocessed input data for SCEG-HiC.calculateHiCWeights and
Run_SCEG_HiCtakes as input the generated data.This function can generate clustered data
or non clustered data for a specific cell type."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-11"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-11');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SCEGHiC</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/pbmc_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregrate</a></li>
    <li><a class="dropdown-item" href="../articles/pbmc_multiomic_CD4T.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</a></li>
    <li><a class="dropdown-item" href="../articles/mouse_skin_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of mouse skin-aggregate</a></li>
    <li><a class="dropdown-item" href="../articles/human_covid_19_scATAC_seq_mono.html">SCEG-HiC on only scATAC-seq data of human COVID 19-Mono</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC/discussions">Discussions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create SCEG-HiC preprocessed input data</h1>
      <small class="dont-index">Source: <a href="https://github.com/wuwei77lx/SCEGHiC/blob/HEAD/R/preprocessing.R" class="external-link"><code>R/preprocessing.R</code></a></small>
      <div class="d-none name"><code>process_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Function to create preprocessed input data for SCEG-HiC.<code>calculateHiCWeights</code> and
<code>Run_SCEG_HiC</code>takes as input the generated data.This function can generate clustered data
or non clustered data for a specific cell type.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">process_data</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  aggregate <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  celltype <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rna_assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  atac_assay <span class="op">=</span> <span class="st">"peaks"</span>,</span>
<span>  cellnames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  atacbinary <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  k_neigh <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  max_overlap <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  size_factor_normalize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  reduction.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">123</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>Seurat object.</p></dd>


<dt id="arg-aggregate">aggregate<a class="anchor" aria-label="anchor" href="#arg-aggregate"></a></dt>
<dd><p>Logically,  should the data be clustered ?If FALSE, <code>celltype</code> must be specified as one cell type.</p></dd>


<dt id="arg-celltype">celltype<a class="anchor" aria-label="anchor" href="#arg-celltype"></a></dt>
<dd><p>Which cells to generate data.</p></dd>


<dt id="arg-rna-assay">rna_assay<a class="anchor" aria-label="anchor" href="#arg-rna-assay"></a></dt>
<dd><p>Name of assay containing gene expression information.If <code>aggregate</code> is TRUE, name of assay containing raw gene expression
information.If <code>aggregate</code> is FALSE, name of assay containing normalize gene expression information.</p></dd>


<dt id="arg-atac-assay">atac_assay<a class="anchor" aria-label="anchor" href="#arg-atac-assay"></a></dt>
<dd><p>Name of assay containing peak information.If <code>aggregate</code> is TRUE, name of assay containing raw peak information.
If <code>aggregate</code> is FALSE, name of assay containing normalize peak information.</p></dd>


<dt id="arg-cellnames">cellnames<a class="anchor" aria-label="anchor" href="#arg-cellnames"></a></dt>
<dd><p>Name of one or more metadata columns to group the cells by. Default is the current cell identities.</p></dd>


<dt id="arg-atacbinary">atacbinary<a class="anchor" aria-label="anchor" href="#arg-atacbinary"></a></dt>
<dd><p>Logical, whether the aggregated scATAC-seq data need binary?</p></dd>


<dt id="arg-k-neigh">k_neigh<a class="anchor" aria-label="anchor" href="#arg-k-neigh"></a></dt>
<dd><p>Number of cells to be aggregated per group.</p></dd>


<dt id="arg-max-overlap">max_overlap<a class="anchor" aria-label="anchor" href="#arg-max-overlap"></a></dt>
<dd><p>The maximum overlapping ratio of two groups.</p></dd>


<dt id="arg-size-factor-normalize">size_factor_normalize<a class="anchor" aria-label="anchor" href="#arg-size-factor-normalize"></a></dt>
<dd><p>Logical, should accessibility values be normalized by size factor?</p></dd>


<dt id="arg-reduction-name">reduction.name<a class="anchor" aria-label="anchor" href="#arg-reduction-name"></a></dt>
<dd><p>The reduction name of extracting the cell coordinates used for aggregating.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, should warning and info messages be printed?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Preprocessed data for SCEG-HiC</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">multiomic_small</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># aggregated with paired scRNA-seq and scATAC-seq data</span></span></span>
<span class="r-in"><span><span class="va">SCEGdata</span> <span class="op">&lt;-</span> <span class="fu">process_data</span><span class="op">(</span><span class="va">multiomic_small</span>, k_neigh <span class="op">=</span> <span class="fl">5</span>, max_overlap <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating aggregated data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Aggregating cluster 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sample cells randomly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> have 11 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Aggregating cluster 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sample cells randomly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> have 11 samples</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># select "1" cells with paired scRNA-seq and scATAC-seq data</span></span></span>
<span class="r-in"><span><span class="va">SCEGdata</span> <span class="op">&lt;-</span> <span class="fu">process_data</span><span class="op">(</span><span class="va">multiomic_small</span>, aggregate <span class="op">=</span> <span class="cn">FALSE</span>, celltype <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xuan Liang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

