<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregate • SCEGHiC</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregate">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-11"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-11');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SCEGHiC</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/pbmc_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregrate</a></li>
    <li><a class="dropdown-item" href="../articles/pbmc_multiomic_CD4T.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</a></li>
    <li><a class="dropdown-item" href="../articles/mouse_skin_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of mouse skin-aggregate</a></li>
    <li><a class="dropdown-item" href="../articles/human_covid_19_scATAC_seq_mono.html">SCEG-HiC on only scATAC-seq data of human COVID 19-Mono</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC/discussions">Discussions</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregate</h1>
                        <h4 data-toc-skip class="author">XuanLiang</h4>
            
            <h4 data-toc-skip class="date">Compiled: February 25,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/wuwei77lx/SCEGHiC/blob/HEAD/vignettes/pbmc_multiomic_aggregate.Rmd" class="external-link"><code>vignettes/pbmc_multiomic_aggregate.Rmd</code></a></small>
      <div class="d-none name"><code>pbmc_multiomic_aggregate.Rmd</code></div>
    </div>

    
    
<p>In this vignette,we’ll demonstrate SCEG-HiC’s capability of inferring
gene-enhancer by applying it to paired scRNA-seq and scATAC-seq data. In
this vignette we’ll be using a publicly available 10x Genomic Multiome
dataset for human PBMCs,and aggregating the data for processing.</p>
<p>The implementent of SCEG-HiC is seamlessly compatible with the
workflow of Seurat/Signac package, SCEG-HiC pipeline consists of the
following three major parts:</p>
<ul>
<li>Set up a Seurat object.</li>
<li>Infer gene-enhancer using SCEG-HiC by taking as input a Seurat
object and the bulk average Hi-C.</li>
<li>Visualize the gene-enhancer.</li>
</ul>
<details><summary><strong>View data download code</strong>
</summary><p>You can download the required data by running the following lines in
a shell:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz.tbi</span></code></pre></div>
</details><div class="section level2">
<h2 id="load-the-required-libraries">Load the required libraries<a class="anchor" aria-label="anchor" href="#load-the-required-libraries"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wuwei77lx/SCEGHiC" class="external-link">SCEGHiC</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://hhoeflin.github.io/hdf5r/" class="external-link">hdf5r</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="set-up-a-seurat-object">Set up a Seurat object<a class="anchor" aria-label="anchor" href="#set-up-a-seurat-object"></a>
</h2>
<p>To facilitate easy exploration, <code>pbmc_multiomic.rds</code> also
available from <a href="https://zenodo.org/record/14849886" class="external-link">10.5281/zenodo.14849886</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the RNA and ATAC data</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="st">"pbmc_granulocyte_sorted_10k_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="va">fragpath</span> <span class="op">&lt;-</span> <span class="st">"pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz"</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get gene annotations for hg38</span></span>
<span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/GetGRangesFromEnsDb.html" class="external-link">GetGRangesFromEnsDb</a></span><span class="op">(</span>ensdb <span class="op">=</span> <span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqlevels</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a Seurat object containing the RNA data</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span><span class="op">$</span><span class="va">`Gene Expression`</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create ATAC assay and add it to the object</span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"ATAC"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CreateChromatinAssay.html" class="external-link">CreateChromatinAssay</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">counts</span><span class="op">$</span><span class="va">Peaks</span>,</span>
<span>  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span>,</span>
<span>  fragments <span class="op">=</span> <span class="va">fragpath</span>,</span>
<span>  annotation <span class="op">=</span> <span class="va">annotation</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 144978 features across 11909 samples within 2 assays </span></span>
<span><span class="co">## Active assay: RNA (36601 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span>
<span><span class="co">##  1 other assay present: ATAC</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Quality control</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/NucleosomeSignal.html" class="external-link">NucleosomeSignal</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/TSSEnrichment.html" class="external-link">TSSEnrichment</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCount_RNA"</span>, <span class="st">"nCount_ATAC"</span>, <span class="st">"TSS.enrichment"</span>, <span class="st">"nucleosome_signal"</span><span class="op">)</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  pt.size <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc_multiomic_aggregate_files/figure-html/unnamed-chunk-6-1.png" width="1728"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter out low quality cells</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  subset <span class="op">=</span> <span class="va">nCount_ATAC</span> <span class="op">&lt;</span> <span class="fl">100000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_RNA</span> <span class="op">&lt;</span> <span class="fl">25000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_ATAC</span> <span class="op">&gt;</span> <span class="fl">1800</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nCount_RNA</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span></span>
<span>    <span class="va">nucleosome_signal</span> <span class="op">&lt;</span> <span class="fl">2</span> <span class="op">&amp;</span></span>
<span>    <span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 144978 features across 11070 samples within 2 assays </span></span>
<span><span class="co">## Active assay: ATAC (108377 features, 0 variable features)</span></span>
<span><span class="co">##  2 layers present: counts, data</span></span>
<span><span class="co">##  1 other assay present: RNA</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># call peaks using MACS2</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CallPeaks.html" class="external-link">CallPeaks</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="co"># remove peaks on nonstandard chromosomes and in genomic blacklist regions</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevels-wrappers.html" class="external-link">keepStandardChromosomes</a></span><span class="op">(</span><span class="va">peaks</span>, pruning.mode <span class="op">=</span> <span class="st">"coarse"</span><span class="op">)</span></span>
<span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">peaks</span>, ranges <span class="op">=</span> <span class="va">blacklist_hg38_unified</span>, invert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># quantify counts in each peak</span></span>
<span><span class="va">macs2_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FeatureMatrix.html" class="external-link">FeatureMatrix</a></span><span class="op">(</span></span>
<span>  fragments <span class="op">=</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/Fragments.html" class="external-link">Fragments</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,</span>
<span>  features <span class="op">=</span> <span class="va">peaks</span>,</span>
<span>  cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Extracting reads overlapping genomic regions</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a new assay using the MACS2 peak set and add it to the Seurat object</span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"peaks"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/CreateChromatinAssay.html" class="external-link">CreateChromatinAssay</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">macs2_counts</span>,</span>
<span>  fragments <span class="op">=</span> <span class="va">fragpath</span>,</span>
<span>  annotation <span class="op">=</span> <span class="va">annotation</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Computing hash</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 276342 features across 11070 samples within 3 assays </span></span>
<span><span class="co">## Active assay: ATAC (108377 features, 0 variable features)</span></span>
<span><span class="co">##  2 layers present: counts, data</span></span>
<span><span class="co">##  2 other assays present: RNA, peaks</span></span></code></pre>
<div class="section level3">
<h3 id="rna-analysis">RNA analysis<a class="anchor" aria-label="anchor" href="#rna-analysis"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"RNA"</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="atac-analysis">ATAC analysis<a class="anchor" aria-label="anchor" href="#atac-analysis"></a>
</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"ATAC"</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FindTopFeatures.html" class="external-link">FindTopFeatures</a></span><span class="op">(</span><span class="va">pbmc</span>, min.cutoff <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunTFIDF.html" class="external-link">RunTFIDF</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunSVD.html" class="external-link">RunSVD</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="annotating-cell-types-and-intergration-analysis">Annotating cell types and Intergration analysis<a class="anchor" aria-label="anchor" href="#annotating-cell-types-and-intergration-analysis"></a>
</h3>
<p>We’ll use an annotated PBMC reference dataset from <a href="https://www.biorxiv.org/content/10.1101/2020.10.12.335331v1" class="external-link">Hao
et al. (2020)</a>, available for download here: <a href="https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat" class="external-link uri">https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat</a>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mojaveazure.github.io/seurat-disk/" class="external-link">SeuratDisk</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load PBMC reference</span></span>
<span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/LoadH5Seurat.html" class="external-link">LoadH5Seurat</a></span><span class="op">(</span><span class="st">"pbmc_multimodal.h5seurat"</span>, assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"SCT"</span> <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span>, reductions <span class="op">=</span> <span class="st">"spca"</span><span class="op">)</span></span>
<span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">reference</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span></span>
<span></span>
<span><span class="co"># transfer cell type labels from reference to query</span></span>
<span><span class="va">transfer_anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindTransferAnchors.html" class="external-link">FindTransferAnchors</a></span><span class="op">(</span></span>
<span>  reference <span class="op">=</span> <span class="va">reference</span>,</span>
<span>  query <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  normalization.method <span class="op">=</span> <span class="st">"SCT"</span>,</span>
<span>  reference.reduction <span class="op">=</span> <span class="st">"spca"</span>,</span>
<span>  recompute.residuals <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">predictions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/TransferData.html" class="external-link">TransferData</a></span><span class="op">(</span></span>
<span>  anchorset <span class="op">=</span> <span class="va">transfer_anchors</span>,</span>
<span>  refdata <span class="op">=</span> <span class="va">reference</span><span class="op">$</span><span class="va">celltype.l1</span>,</span>
<span>  weight.reduction <span class="op">=</span> <span class="va">pbmc</span><span class="op">[[</span><span class="st">"pca"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AddMetaData.html" class="external-link">AddMetaData</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">predictions</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set the cell identities to the cell type predictions</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"predicted.id"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4 T"</span>, <span class="st">"CD8 T"</span>, <span class="st">"Mono"</span>, <span class="st">"NK"</span>, <span class="st">"B"</span>, <span class="st">"DC"</span>, <span class="st">"other T"</span>, <span class="st">"other"</span><span class="op">)</span></span></code></pre></div>
<p>Using the weighted nearest neighbor methods in <a href="https://www.biorxiv.org/content/10.1101/2020.10.12.335331v1" class="external-link">Seurat
v4</a>, we can compute a joint neighbor graph that represent both the
gene expression and DNA accessibility measurements.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build a joint neighbor graph using both assays</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMultiModalNeighbors.html" class="external-link">FindMultiModalNeighbors</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  reduction.list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"pca"</span>, <span class="st">"lsi"</span><span class="op">)</span>,</span>
<span>  dims.list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">40</span><span class="op">)</span>,</span>
<span>  modality.weight.name <span class="op">=</span> <span class="st">"RNA.weight"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># build a joint UMAP visualization</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>  nn.name <span class="op">=</span> <span class="st">"weighted.nn"</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, repel <span class="op">=</span> <span class="cn">TRUE</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc_multiomic_aggregate_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="infer-gene-enhancer-using-sceg-hic">Infer gene-enhancer using SCEG-HiC<a class="anchor" aria-label="anchor" href="#infer-gene-enhancer-using-sceg-hic"></a>
</h2>
<div class="section level3">
<h3 id="preprocess-data">Preprocess data<a class="anchor" aria-label="anchor" href="#preprocess-data"></a>
</h3>
<p>When preprocessing data with SCEG-HiC, you can choose between
clustering or single-cell type analysis. By default, the scATAC-seq
counts are binarized.In addition, you can select either paired scRNA-seq
and scATAC-seq data or only scATAC-seq data. Here, we have aggregated
with paired scRNA-seq and scATAC-seq data.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SCEGdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">pbmc</span>, max_overlap <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Generating aggregated data</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster CD4 T</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 73 samples</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster CD8 T</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 41 samples</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster Mono</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 79 samples</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster NK</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 12 samples</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster B</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 18 samples</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster DC</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 5 samples</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster other T</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 6 samples</span></span></code></pre>
<pre><code><span><span class="co">## Aggregating cluster other</span></span></code></pre>
<pre><code><span><span class="co">## Sample cells randomly.</span></span></code></pre>
<pre><code><span><span class="co">## have 2 samples</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="calculate-weight">Calculate weight<a class="anchor" aria-label="anchor" href="#calculate-weight"></a>
</h3>
<p>SCEG-HiC takes a long time to calculate weight, and the process
becomes slower as more genes are selected. SCEG-HiC provides download
websites for the average Hi-C data for both humans and mice. In this
case, we select the human average Hi-C data for analysis.</p>
<details><summary><strong>Select highly variable features</strong>
</summary><p>You can select 3000 highly variable features:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">assays</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">var.features</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHiCWeights.html">calculateHiCWeights</a></span><span class="op">(</span><span class="va">SCEGdata</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="va">gene</span>, averHicPath <span class="op">=</span> <span class="st">"/picb/bigdata/project/liangxuan/data/human_contact/AvgHiC"</span><span class="op">)</span></span></code></pre></div>
</details><div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For example, Calculate weight for CD8 T's markers: PDK1</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHiCWeights.html">calculateHiCWeights</a></span><span class="op">(</span><span class="va">SCEGdata</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="st">"PDK1"</span>, averHicPath <span class="op">=</span> <span class="st">"/picb/bigdata/project/liangxuan/data/human_contact/AvgHiC"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Successfully obtained 1 TSS loci of genes from chromosome 2.</span></span></code></pre>
<pre><code><span><span class="co">## Calculate weight of  PDK1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-model">Run model<a class="anchor" aria-label="anchor" href="#run-model"></a>
</h3>
<p>SCEG-HiC uses wglasso machine learning method and normalizes the bulk
average Hi-C matrix using rank score.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For example, Run model for CD8 T's markers: PDK1</span></span>
<span><span class="va">results_SCEGHiC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Run_SCEG_HiC.html">Run_SCEG_HiC</a></span><span class="op">(</span><span class="va">SCEGdata</span>, <span class="va">weight</span>, focus_gene <span class="op">=</span> <span class="st">"PDK1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The predicted genes are 1 in total.</span></span></code></pre>
<pre><code><span><span class="co">## Run model of  PDK1</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="visualize-the-gene-enhancer">Visualize the gene-enhancer<a class="anchor" aria-label="anchor" href="#visualize-the-gene-enhancer"></a>
</h2>
<div class="section level3">
<h3 id="visualize-the-links">Visualize the links<a class="anchor" aria-label="anchor" href="#visualize-the-links"></a>
</h3>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://hgdownload2.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.refGene.gtf.gz"</span>, <span class="va">temp</span><span class="op">)</span></span>
<span><span class="va">gene_anno</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/readGFF.html" class="external-link">readGFF</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span></span>
<span><span class="co"># rename some columns to match requirements</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">chromosome</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">seqid</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">gene_id</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">transcript</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">transcript_id</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">symbol</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">gene_name</span></span>
<span><span class="fu"><a href="../reference/connections_Plot.html">connections_Plot</a></span><span class="op">(</span><span class="va">results_SCEGHiC</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="st">"PDK1"</span>, cutoff <span class="op">=</span> <span class="fl">0.01</span>, gene_anno <span class="op">=</span> <span class="va">gene_anno</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc_multiomic_aggregate_files/figure-html/unnamed-chunk-18-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="coverage-plot-and-visualize-the-links-of-pdk1">Coverage plot and visualize the links of PDK1<a class="anchor" aria-label="anchor" href="#coverage-plot-and-visualize-the-links-of-pdk1"></a>
</h3>
<p>Here, the truth of CD8 T is sourced from the <a href="https://www.encodeproject.org/files/ENCFF063KHP/@@download/ENCFF063KHP.hic" class="external-link">ENCODE
database</a>, and the correlation is calculated by computing the Pearson
correlation coefficient between each gene and enhancer.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the truth of CD8 T data</span></span>
<span><span class="va">truth_CD8</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"truth_CD8.rds"</span><span class="op">)</span></span>
<span><span class="co">## load correlation data</span></span>
<span><span class="va">correlation</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"correlation_CD8.rds"</span><span class="op">)</span></span>
<span><span class="co"># Coverage plot and visualize the links of PDK1</span></span>
<span><span class="fu"><a href="../reference/coverPlot.html">coverPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, focus_gene <span class="op">=</span> <span class="st">"PDK1"</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, assay <span class="op">=</span> <span class="st">"peaks"</span>, HIC_Result <span class="op">=</span> <span class="va">truth_CD8</span>, SCEG_HiC_Result <span class="op">=</span> <span class="va">results_SCEGHiC</span>, SCEG_HiC_cutoff <span class="op">=</span> <span class="fl">0.01</span>, correlation <span class="op">=</span> <span class="va">correlation</span>, cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4 T"</span>, <span class="st">"CD8 T"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc_multiomic_aggregate_files/figure-html/unnamed-chunk-19-1.png" width="1488"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-conda-linux-gnu</span></span>
<span><span class="co">## Running under: CentOS Linux 7 (Core)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /home/liangxuan/conda/envs/test/lib/libopenblasp-r0.3.28.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Asia/Shanghai</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices datasets  utils     methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] SeuratDisk_0.0.0.9021             hdf5r_1.3.12                     </span></span>
<span><span class="co">##  [3] ggplot2_3.5.1                     BSgenome.Hsapiens.UCSC.hg38_1.4.5</span></span>
<span><span class="co">##  [5] BSgenome_1.74.0                   rtracklayer_1.66.0               </span></span>
<span><span class="co">##  [7] BiocIO_1.16.0                     Biostrings_2.74.1                </span></span>
<span><span class="co">##  [9] XVector_0.46.0                    EnsDb.Hsapiens.v86_2.99.0        </span></span>
<span><span class="co">## [11] ensembldb_2.30.0                  AnnotationFilter_1.30.0          </span></span>
<span><span class="co">## [13] GenomicFeatures_1.58.0            AnnotationDbi_1.68.0             </span></span>
<span><span class="co">## [15] Biobase_2.66.0                    GenomicRanges_1.58.0             </span></span>
<span><span class="co">## [17] GenomeInfoDb_1.42.1               IRanges_2.40.1                   </span></span>
<span><span class="co">## [19] S4Vectors_0.44.0                  BiocGenerics_0.52.0              </span></span>
<span><span class="co">## [21] Seurat_5.2.0                      SeuratObject_5.0.2               </span></span>
<span><span class="co">## [23] sp_2.1-4                          Signac_1.14.9001                 </span></span>
<span><span class="co">## [25] SCEGHiC_0.0.0.9000               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.5                    ProtGenerics_1.38.0        </span></span>
<span><span class="co">##   [3] matrixStats_1.5.0           spatstat.sparse_3.1-0      </span></span>
<span><span class="co">##   [5] bitops_1.0-9                httr_1.4.7                 </span></span>
<span><span class="co">##   [7] RColorBrewer_1.1-3          tools_4.4.2                </span></span>
<span><span class="co">##   [9] sctransform_0.4.1           backports_1.5.0            </span></span>
<span><span class="co">##  [11] R6_2.5.1                    uwot_0.2.2                 </span></span>
<span><span class="co">##  [13] lazyeval_0.2.2              Gviz_1.50.0                </span></span>
<span><span class="co">##  [15] cicero_1.3.9                withr_3.0.2                </span></span>
<span><span class="co">##  [17] prettyunits_1.2.0           gridExtra_2.3              </span></span>
<span><span class="co">##  [19] progressr_0.15.1            cli_3.6.3                  </span></span>
<span><span class="co">##  [21] textshaping_0.4.0           spatstat.explore_3.3-4     </span></span>
<span><span class="co">##  [23] fastDummies_1.7.4           labeling_0.4.3             </span></span>
<span><span class="co">##  [25] slam_0.1-55                 sass_0.4.9                 </span></span>
<span><span class="co">##  [27] spatstat.data_3.1-4         ggridges_0.5.6             </span></span>
<span><span class="co">##  [29] pbapply_1.7-2               pkgdown_2.1.1              </span></span>
<span><span class="co">##  [31] Rsamtools_2.22.0            systemfonts_1.1.0          </span></span>
<span><span class="co">##  [33] foreign_0.8-88              R.utils_2.12.3             </span></span>
<span><span class="co">##  [35] dichromat_2.0-0.1           parallelly_1.41.0          </span></span>
<span><span class="co">##  [37] VGAM_1.1-12                 rstudioapi_0.17.1          </span></span>
<span><span class="co">##  [39] RSQLite_2.3.9               FNN_1.1.4.1                </span></span>
<span><span class="co">##  [41] generics_0.1.3              ica_1.0-3                  </span></span>
<span><span class="co">##  [43] spatstat.random_3.3-2       dplyr_1.1.4                </span></span>
<span><span class="co">##  [45] Matrix_1.6-5                interp_1.1-6               </span></span>
<span><span class="co">##  [47] ggbeeswarm_0.7.2            abind_1.4-8                </span></span>
<span><span class="co">##  [49] R.methodsS3_1.8.2           lifecycle_1.0.4            </span></span>
<span><span class="co">##  [51] yaml_2.3.10                 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [53] SparseArray_1.6.0           BiocFileCache_2.14.0       </span></span>
<span><span class="co">##  [55] Rtsne_0.17                  grid_4.4.2                 </span></span>
<span><span class="co">##  [57] blob_1.2.4                  promises_1.3.2             </span></span>
<span><span class="co">##  [59] crayon_1.5.3                miniUI_0.1.1.1             </span></span>
<span><span class="co">##  [61] lattice_0.22-6              cowplot_1.1.3              </span></span>
<span><span class="co">##  [63] KEGGREST_1.46.0             pillar_1.10.1              </span></span>
<span><span class="co">##  [65] knitr_1.49                  boot_1.3-31                </span></span>
<span><span class="co">##  [67] rjson_0.2.23                future.apply_1.11.3        </span></span>
<span><span class="co">##  [69] codetools_0.2-20            fastmatch_1.1-6            </span></span>
<span><span class="co">##  [71] glue_1.8.0                  spatstat.univar_3.1-1      </span></span>
<span><span class="co">##  [73] data.table_1.16.4           Rdpack_2.6.2               </span></span>
<span><span class="co">##  [75] vctrs_0.6.5                 png_0.1-8                  </span></span>
<span><span class="co">##  [77] spam_2.11-0                 gtable_0.3.6               </span></span>
<span><span class="co">##  [79] assertthat_0.2.1            cachem_1.1.0               </span></span>
<span><span class="co">##  [81] xfun_0.50                   rbibutils_2.3              </span></span>
<span><span class="co">##  [83] S4Arrays_1.6.0              mime_0.12                  </span></span>
<span><span class="co">##  [85] reformulas_0.4.0            survival_3.8-3             </span></span>
<span><span class="co">##  [87] SingleCellExperiment_1.28.1 RcppRoll_0.3.1             </span></span>
<span><span class="co">##  [89] fitdistrplus_1.2-2          ROCR_1.0-11                </span></span>
<span><span class="co">##  [91] nlme_3.1-166                bit64_4.5.2                </span></span>
<span><span class="co">##  [93] progress_1.2.3              filelock_1.0.3             </span></span>
<span><span class="co">##  [95] RcppAnnoy_0.0.22            bslib_0.8.0                </span></span>
<span><span class="co">##  [97] irlba_2.3.5.1               vipor_0.4.7                </span></span>
<span><span class="co">##  [99] KernSmooth_2.23-26          rpart_4.1.24               </span></span>
<span><span class="co">## [101] colorspace_2.1-1            DBI_1.2.3                  </span></span>
<span><span class="co">## [103] Hmisc_5.2-2                 nnet_7.3-20                </span></span>
<span><span class="co">## [105] ggrastr_1.0.2               tidyselect_1.2.1           </span></span>
<span><span class="co">## [107] bit_4.5.0.1                 compiler_4.4.2             </span></span>
<span><span class="co">## [109] curl_6.0.1                  httr2_1.0.7                </span></span>
<span><span class="co">## [111] htmlTable_2.4.3             xml2_1.3.6                 </span></span>
<span><span class="co">## [113] desc_1.4.3                  DelayedArray_0.32.0        </span></span>
<span><span class="co">## [115] plotly_4.10.4               checkmate_2.3.2            </span></span>
<span><span class="co">## [117] scales_1.3.0                lmtest_0.9-40              </span></span>
<span><span class="co">## [119] rappdirs_0.3.3              goftest_1.2-3              </span></span>
<span><span class="co">## [121] stringr_1.5.1               digest_0.6.37              </span></span>
<span><span class="co">## [123] minqa_1.2.8                 spatstat.utils_3.1-2       </span></span>
<span><span class="co">## [125] reader_1.0.6                rmarkdown_2.29             </span></span>
<span><span class="co">## [127] htmltools_0.5.8.1           pkgconfig_2.0.3            </span></span>
<span><span class="co">## [129] jpeg_0.1-10                 base64enc_0.1-3            </span></span>
<span><span class="co">## [131] lme4_1.1-36                 MatrixGenerics_1.18.1      </span></span>
<span><span class="co">## [133] dbplyr_2.5.0                fastmap_1.2.0              </span></span>
<span><span class="co">## [135] rlang_1.1.4                 htmlwidgets_1.6.4          </span></span>
<span><span class="co">## [137] UCSC.utils_1.2.0            shiny_1.10.0               </span></span>
<span><span class="co">## [139] farver_2.1.2                jquerylib_0.1.4            </span></span>
<span><span class="co">## [141] zoo_1.8-12                  jsonlite_1.8.9             </span></span>
<span><span class="co">## [143] BiocParallel_1.40.0         R.oo_1.27.0                </span></span>
<span><span class="co">## [145] VariantAnnotation_1.52.0    RCurl_1.98-1.16            </span></span>
<span><span class="co">## [147] magrittr_2.0.3              Formula_1.2-5              </span></span>
<span><span class="co">## [149] GenomeInfoDbData_1.2.13     dotCall64_1.2              </span></span>
<span><span class="co">## [151] patchwork_1.3.0             munsell_0.5.1              </span></span>
<span><span class="co">## [153] Rcpp_1.0.14                 reticulate_1.40.0          </span></span>
<span><span class="co">## [155] stringi_1.8.4               zlibbioc_1.52.0            </span></span>
<span><span class="co">## [157] MASS_7.3-64                 plyr_1.8.9                 </span></span>
<span><span class="co">## [159] parallel_4.4.2              listenv_0.9.1              </span></span>
<span><span class="co">## [161] ggrepel_0.9.6               deldir_2.0-4               </span></span>
<span><span class="co">## [163] splines_4.4.2               tensor_1.5                 </span></span>
<span><span class="co">## [165] hms_1.1.3                   igraph_2.0.3               </span></span>
<span><span class="co">## [167] spatstat.geom_3.3-4         RcppHNSW_0.6.0             </span></span>
<span><span class="co">## [169] reshape2_1.4.4              biomaRt_2.62.0             </span></span>
<span><span class="co">## [171] XML_3.99-0.17               evaluate_1.0.3             </span></span>
<span><span class="co">## [173] latticeExtra_0.6-30         biovizBase_1.54.0          </span></span>
<span><span class="co">## [175] renv_1.0.11                 BiocManager_1.30.25        </span></span>
<span><span class="co">## [177] NCmisc_1.2.0                nloptr_2.1.1               </span></span>
<span><span class="co">## [179] tweenr_2.0.3                httpuv_1.6.15              </span></span>
<span><span class="co">## [181] RANN_2.6.2                  tidyr_1.3.1                </span></span>
<span><span class="co">## [183] purrr_1.0.2                 polyclip_1.10-7            </span></span>
<span><span class="co">## [185] future_1.34.0               scattermore_1.2            </span></span>
<span><span class="co">## [187] ggforce_0.4.2               monocle3_1.3.7             </span></span>
<span><span class="co">## [189] xtable_1.8-4                restfulr_0.0.15            </span></span>
<span><span class="co">## [191] RSpectra_0.16-2             later_1.4.1                </span></span>
<span><span class="co">## [193] glasso_1.11                 viridisLite_0.4.2          </span></span>
<span><span class="co">## [195] ragg_1.3.3                  tibble_3.2.1               </span></span>
<span><span class="co">## [197] beeswarm_0.4.0              memoise_2.0.1              </span></span>
<span><span class="co">## [199] GenomicAlignments_1.42.0    cluster_2.1.8              </span></span>
<span><span class="co">## [201] globals_0.16.3</span></span></code></pre>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xuan Liang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
