<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run SCEG-HiC's model — Run_SCEG_HiC • SCEGHiC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run SCEG-HiC's model — Run_SCEG_HiC"><meta name="description" content="The model uses data preprocessed by SCEG-HiC, the bulk average Hi-C as the penalty matrix,
calculates the partial correlation coefficient matrix between elements using the wglasso model,
and selects the gene-peak."><meta property="og:description" content="The model uses data preprocessed by SCEG-HiC, the bulk average Hi-C as the penalty matrix,
calculates the partial correlation coefficient matrix between elements using the wglasso model,
and selects the gene-peak."><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-11"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-11');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SCEGHiC</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/pbmc_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregrate</a></li>
    <li><a class="dropdown-item" href="../articles/pbmc_multiomic_CD4T.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</a></li>
    <li><a class="dropdown-item" href="../articles/mouse_skin_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of mouse skin-aggregate</a></li>
    <li><a class="dropdown-item" href="../articles/human_covid_19_scATAC_seq_mono.html">SCEG-HiC on only scATAC-seq data of human COVID 19-Mono</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC/discussions">Discussions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Run SCEG-HiC's model</h1>
      <small class="dont-index">Source: <a href="https://github.com/wuwei77lx/SCEGHiC/blob/HEAD/R/model.R" class="external-link"><code>R/model.R</code></a></small>
      <div class="d-none name"><code>Run_SCEG_HiC.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The model uses data preprocessed by SCEG-HiC, the bulk average Hi-C as the penalty matrix,
calculates the partial correlation coefficient matrix between elements using the wglasso model,
and selects the gene-peak.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Run_SCEG_HiC</span><span class="op">(</span></span>
<span>  <span class="va">SCEGdata</span>,</span>
<span>  <span class="va">HiCWeights</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wglasso"</span>,</span>
<span>  <span class="va">focus_gene</span>,</span>
<span>  normalizeMethod <span class="op">=</span> <span class="st">"1"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-scegdata">SCEGdata<a class="anchor" aria-label="anchor" href="#arg-scegdata"></a></dt>
<dd><p>Preprocessed data for SCEG-HiC.</p></dd>


<dt id="arg-hicweights">HiCWeights<a class="anchor" aria-label="anchor" href="#arg-hicweights"></a></dt>
<dd><p>Raw the bulk average Hi-C matrix.Run the result of code'calculateHiCWeights'.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>The method of gene network reconstruction.Default is wglasso.</p></dd>


<dt id="arg-focus-gene">focus_gene<a class="anchor" aria-label="anchor" href="#arg-focus-gene"></a></dt>
<dd><p>The focused genes.</p></dd>


<dt id="arg-normalizemethod">normalizeMethod<a class="anchor" aria-label="anchor" href="#arg-normalizemethod"></a></dt>
<dd><p>The method of normalizing the average Hi-C.Default is "1".</p><ul><li><p><code>'1'</code>,rank score to normalize the average Hi-C.</p></li>
<li><p><code>'2'</code>,min-max to normalize the average Hi-C.</p></li>
<li><p><code>'3'</code>,0-1 to normalize the average Hi-C.</p></li>
<li><p><code>'4'</code>,-log10 to normalize the average Hi-C.</p></li>
</ul></dd>


<dt id="arg-cutoff">cutoff<a class="anchor" aria-label="anchor" href="#arg-cutoff"></a></dt>
<dd><p>The threshold of gene-peak.Default is NULL.If <code>"aggregate"</code> is TRUE ,we recommend the cutoff of 0.01.
If <code>"aggregate"</code> is FALSE ,we recommend the cutoff of 0.001.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, should warning and info messages be printed?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data.frame of correlations for each gene-peak pair.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">multiomic_small</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">SCEGdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="process_data.html">process_data</a></span><span class="op">(</span><span class="va">multiomic_small</span>, k_neigh <span class="op">=</span> <span class="fl">5</span>, max_overlap <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating aggregated data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Aggregating cluster 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sample cells randomly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> have 11 samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Aggregating cluster 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sample cells randomly.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> have 11 samples</span>
<span class="r-in"><span><span class="va">fpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"SCEGHiC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TRABD2A"</span>, <span class="st">"GNLY"</span>, <span class="st">"MFSD6"</span>, <span class="st">"CTLA4"</span>, <span class="st">"LCLAT1"</span>, <span class="st">"NCK2"</span>, <span class="st">"GALM"</span>, <span class="st">"TMSB10"</span>, <span class="st">"ID2"</span>, <span class="st">"CXCR4"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="calculateHiCWeights.html">calculateHiCWeights</a></span><span class="op">(</span><span class="va">SCEGdata</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="va">gene</span>, averHicPath <span class="op">=</span> <span class="va">fpath</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Successfully obtained 10 TSS loci of genes from chromosome 2.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  TRABD2A</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  GNLY</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  MFSD6</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  CXCR4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  CTLA4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  LCLAT1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  NCK2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  ID2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  GALM</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculate weight of  TMSB10</span>
<span class="r-in"><span><span class="va">results_SCEGHiC</span> <span class="op">&lt;-</span> <span class="fu">Run_SCEG_HiC</span><span class="op">(</span><span class="va">SCEGdata</span>, <span class="va">weight</span>, focus_gene <span class="op">=</span> <span class="va">gene</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The predicted genes are 10 in total.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  TRABD2A</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  GNLY</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  MFSD6</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  CXCR4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  CTLA4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  LCLAT1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  NCK2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  ID2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  GALM</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Run model of  TMSB10</span>
<span class="r-in"><span><span class="va">results_SCEGHiC</span><span class="op">[</span><span class="va">results_SCEGHiC</span><span class="op">$</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"CTLA4"</span>, <span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           gene                     peak       score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2_203623664_203623982 CTLA4 chr2_203623664_203623982  0.03623216</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2_203730093_203731243 CTLA4 chr2_203730093_203731243  0.14814205</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2_203786816_203787404 CTLA4 chr2_203786816_203787404  0.00000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2_203932737_203935679 CTLA4 chr2_203932737_203935679 -0.10013637</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2_203945543_203945939 CTLA4 chr2_203945543_203945939  0.00000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2_203992800_203993979 CTLA4 chr2_203992800_203993979  0.43240254</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr2_204109915_204110871 CTLA4 chr2_204109915_204110871  0.00000000</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xuan Liang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

