<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T • SCEGHiC</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-119440867-11"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-119440867-11');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SCEGHiC</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/pbmc_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-aggregrate</a></li>
    <li><a class="dropdown-item" href="../articles/pbmc_multiomic_CD4T.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</a></li>
    <li><a class="dropdown-item" href="../articles/mouse_skin_multiomic_aggregate.html">SCEG-HiC on paired scRNA-seq and scATAC-seq data of mouse skin-aggregate</a></li>
    <li><a class="dropdown-item" href="../articles/human_covid_19_scATAC_seq_mono.html">SCEG-HiC on only scATAC-seq data of human COVID 19-Mono</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC/discussions">Discussions</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/wuwei77lx/SCEGHiC"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SCEG-HiC on paired scRNA-seq and scATAC-seq data of PBMC-CD4T</h1>
                        <h4 data-toc-skip class="author">XuanLiang</h4>
            
            <h4 data-toc-skip class="date">Compiled: February 25,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/wuwei77lx/SCEGHiC/blob/HEAD/vignettes/pbmc_multiomic_CD4T.Rmd" class="external-link"><code>vignettes/pbmc_multiomic_CD4T.Rmd</code></a></small>
      <div class="d-none name"><code>pbmc_multiomic_CD4T.Rmd</code></div>
    </div>

    
    
<p>In this vignette,we’ll demonstrate SCEG-HiC’s capability of inferring
gene-enhancer by applying it to paired scRNA-seq and scATAC-seq data. In
this vignette we’ll be using a publicly available 10x Genomic Multiome
dataset for human PBMCs,and only select CD4 T cells for processing</p>
<p>Ok, so let’s get started. First, we need a Seurat object that has
paired scRNA-seq and scATAC-seq data, respectively. Here we use multiome
data from human PBMCs, processed as demonstrated in the <a href="https://wuwei77lx.github.io/SCEGHiC/articles/pbmc_multiomic_aggregrate.html">SCEG-HiC
on paired scRNA-seq and scATAC-seq data of
PBMC-aggregate</a>,<code>pbmc_multiomic.rds</code> also available from
<a href="https://zenodo.org/record/14849886" class="external-link">10.5281/zenodo.14849886</a>.</p>
<p>The implementent of SCEG-HiC is seamlessly compatible with the
workflow of Seurat/Signac package, SCEG-HiC pipeline consists of the
following two major parts:</p>
<ul>
<li>Infer gene-enhancer using SCEG-HiC by taking as input a Seurat
object and the bulk average Hi-C.</li>
<li>Visualize the gene-enhancer.</li>
</ul>
<div class="section level2">
<h2 id="load-the-required-libraries">Load the required libraries<a class="anchor" aria-label="anchor" href="#load-the-required-libraries"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wuwei77lx/SCEGHiC" class="external-link">SCEGHiC</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="infer-gene-enhancer-using-sceg-hic">Infer gene-enhancer using SCEG-HiC<a class="anchor" aria-label="anchor" href="#infer-gene-enhancer-using-sceg-hic"></a>
</h2>
<div class="section level3">
<h3 id="preprocess-data">Preprocess data<a class="anchor" aria-label="anchor" href="#preprocess-data"></a>
</h3>
<p>Here, we only select CD8 T cells with paired scRNA-seq and scATAC-seq
data.By default, the scATAC-seq counts are binarized.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load data</span></span>
<span><span class="va">pbmc</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"pbmc_multiomic.rds"</span><span class="op">)</span></span>
<span><span class="va">SCEGdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_data.html">process_data</a></span><span class="op">(</span><span class="va">pbmc</span>, aggregate <span class="op">=</span> <span class="cn">FALSE</span>, celltype <span class="op">=</span> <span class="st">"CD4 T"</span>, rna_assay <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculate-weight">Calculate weight<a class="anchor" aria-label="anchor" href="#calculate-weight"></a>
</h3>
<p>SCEG-HiC takes a long time to calculate weight, and the process
becomes slower as more genes are selected. SCEG-HiC provides download
websites for the average Hi-C data for both humans and mice. In this
case, we select the human average Hi-C data for analysis.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For example, Calculate weight for CD4 T's markers: GLB1</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHiCWeights.html">calculateHiCWeights</a></span><span class="op">(</span><span class="va">SCEGdata</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="st">"GLB1"</span>, averHicPath <span class="op">=</span> <span class="st">"/picb/bigdata/project/liangxuan/data/human_contact/AvgHiC"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Successfully obtained 1 TSS loci of genes from chromosome 3.</span></span></code></pre>
<pre><code><span><span class="co">## Calculate weight of  GLB1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-model">Run model<a class="anchor" aria-label="anchor" href="#run-model"></a>
</h3>
<p>SCEG-HiC uses wglasso machine learning method and normalizes the bulk
average Hi-C matrix using rank score.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For example, Run model for CD4 T's markers: GLB1</span></span>
<span><span class="va">results_SCEGHiC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Run_SCEG_HiC.html">Run_SCEG_HiC</a></span><span class="op">(</span><span class="va">SCEGdata</span>, <span class="va">weight</span>, focus_gene <span class="op">=</span> <span class="st">"GLB1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The predicted genes are 1 in total.</span></span></code></pre>
<pre><code><span><span class="co">## Run model of  GLB1</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="visualize-the-gene-enhancer">Visualize the gene-enhancer<a class="anchor" aria-label="anchor" href="#visualize-the-gene-enhancer"></a>
</h2>
<div class="section level3">
<h3 id="visualize-the-links">Visualize the links<a class="anchor" aria-label="anchor" href="#visualize-the-links"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://hgdownload2.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.refGene.gtf.gz"</span>, <span class="va">temp</span><span class="op">)</span></span>
<span><span class="va">gene_anno</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/readGFF.html" class="external-link">readGFF</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span></span>
<span><span class="co"># rename some columns to match requirements</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">chromosome</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">seqid</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">gene_id</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">transcript</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">transcript_id</span></span>
<span><span class="va">gene_anno</span><span class="op">$</span><span class="va">symbol</span> <span class="op">&lt;-</span> <span class="va">gene_anno</span><span class="op">$</span><span class="va">gene_name</span></span>
<span><span class="fu"><a href="../reference/connections_Plot.html">connections_Plot</a></span><span class="op">(</span><span class="va">results_SCEGHiC</span>, species <span class="op">=</span> <span class="st">"Homo sapiens"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, focus_gene <span class="op">=</span> <span class="st">"GLB1"</span>, cutoff <span class="op">=</span> <span class="fl">0.001</span>, gene_anno <span class="op">=</span> <span class="va">gene_anno</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc_multiomic_CD4T_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-conda-linux-gnu</span></span>
<span><span class="co">## Running under: CentOS Linux 7 (Core)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /home/liangxuan/conda/envs/test/lib/libopenblasp-r0.3.28.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Asia/Shanghai</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices datasets  utils     methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggplot2_3.5.1                     BSgenome.Hsapiens.UCSC.hg38_1.4.5</span></span>
<span><span class="co">##  [3] BSgenome_1.74.0                   rtracklayer_1.66.0               </span></span>
<span><span class="co">##  [5] BiocIO_1.16.0                     Biostrings_2.74.1                </span></span>
<span><span class="co">##  [7] XVector_0.46.0                    EnsDb.Hsapiens.v86_2.99.0        </span></span>
<span><span class="co">##  [9] ensembldb_2.30.0                  AnnotationFilter_1.30.0          </span></span>
<span><span class="co">## [11] GenomicFeatures_1.58.0            AnnotationDbi_1.68.0             </span></span>
<span><span class="co">## [13] Biobase_2.66.0                    GenomicRanges_1.58.0             </span></span>
<span><span class="co">## [15] GenomeInfoDb_1.42.1               IRanges_2.40.1                   </span></span>
<span><span class="co">## [17] S4Vectors_0.44.0                  BiocGenerics_0.52.0              </span></span>
<span><span class="co">## [19] Seurat_5.2.0                      SeuratObject_5.0.2               </span></span>
<span><span class="co">## [21] sp_2.1-4                          Signac_1.14.9001                 </span></span>
<span><span class="co">## [23] SCEGHiC_0.0.0.9000               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.5                    ProtGenerics_1.38.0        </span></span>
<span><span class="co">##   [3] matrixStats_1.5.0           spatstat.sparse_3.1-0      </span></span>
<span><span class="co">##   [5] bitops_1.0-9                httr_1.4.7                 </span></span>
<span><span class="co">##   [7] RColorBrewer_1.1-3          tools_4.4.2                </span></span>
<span><span class="co">##   [9] sctransform_0.4.1           backports_1.5.0            </span></span>
<span><span class="co">##  [11] R6_2.5.1                    uwot_0.2.2                 </span></span>
<span><span class="co">##  [13] lazyeval_0.2.2              Gviz_1.50.0                </span></span>
<span><span class="co">##  [15] cicero_1.3.9                withr_3.0.2                </span></span>
<span><span class="co">##  [17] prettyunits_1.2.0           gridExtra_2.3              </span></span>
<span><span class="co">##  [19] progressr_0.15.1            cli_3.6.3                  </span></span>
<span><span class="co">##  [21] textshaping_0.4.0           spatstat.explore_3.3-4     </span></span>
<span><span class="co">##  [23] fastDummies_1.7.4           slam_0.1-55                </span></span>
<span><span class="co">##  [25] sass_0.4.9                  spatstat.data_3.1-4        </span></span>
<span><span class="co">##  [27] ggridges_0.5.6              pbapply_1.7-2              </span></span>
<span><span class="co">##  [29] pkgdown_2.1.1               Rsamtools_2.22.0           </span></span>
<span><span class="co">##  [31] systemfonts_1.1.0           foreign_0.8-88             </span></span>
<span><span class="co">##  [33] R.utils_2.12.3              dichromat_2.0-0.1          </span></span>
<span><span class="co">##  [35] parallelly_1.41.0           VGAM_1.1-12                </span></span>
<span><span class="co">##  [37] rstudioapi_0.17.1           RSQLite_2.3.9              </span></span>
<span><span class="co">##  [39] FNN_1.1.4.1                 generics_0.1.3             </span></span>
<span><span class="co">##  [41] ica_1.0-3                   spatstat.random_3.3-2      </span></span>
<span><span class="co">##  [43] dplyr_1.1.4                 Matrix_1.6-5               </span></span>
<span><span class="co">##  [45] interp_1.1-6                abind_1.4-8                </span></span>
<span><span class="co">##  [47] R.methodsS3_1.8.2           lifecycle_1.0.4            </span></span>
<span><span class="co">##  [49] yaml_2.3.10                 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [51] SparseArray_1.6.0           BiocFileCache_2.14.0       </span></span>
<span><span class="co">##  [53] Rtsne_0.17                  grid_4.4.2                 </span></span>
<span><span class="co">##  [55] blob_1.2.4                  promises_1.3.2             </span></span>
<span><span class="co">##  [57] crayon_1.5.3                miniUI_0.1.1.1             </span></span>
<span><span class="co">##  [59] lattice_0.22-6              cowplot_1.1.3              </span></span>
<span><span class="co">##  [61] KEGGREST_1.46.0             pillar_1.10.1              </span></span>
<span><span class="co">##  [63] knitr_1.49                  boot_1.3-31                </span></span>
<span><span class="co">##  [65] rjson_0.2.23                future.apply_1.11.3        </span></span>
<span><span class="co">##  [67] codetools_0.2-20            fastmatch_1.1-6            </span></span>
<span><span class="co">##  [69] glue_1.8.0                  spatstat.univar_3.1-1      </span></span>
<span><span class="co">##  [71] data.table_1.16.4           Rdpack_2.6.2               </span></span>
<span><span class="co">##  [73] vctrs_0.6.5                 png_0.1-8                  </span></span>
<span><span class="co">##  [75] spam_2.11-0                 gtable_0.3.6               </span></span>
<span><span class="co">##  [77] assertthat_0.2.1            cachem_1.1.0               </span></span>
<span><span class="co">##  [79] xfun_0.50                   rbibutils_2.3              </span></span>
<span><span class="co">##  [81] S4Arrays_1.6.0              mime_0.12                  </span></span>
<span><span class="co">##  [83] reformulas_0.4.0            survival_3.8-3             </span></span>
<span><span class="co">##  [85] SingleCellExperiment_1.28.1 RcppRoll_0.3.1             </span></span>
<span><span class="co">##  [87] fitdistrplus_1.2-2          ROCR_1.0-11                </span></span>
<span><span class="co">##  [89] nlme_3.1-166                bit64_4.5.2                </span></span>
<span><span class="co">##  [91] progress_1.2.3              filelock_1.0.3             </span></span>
<span><span class="co">##  [93] RcppAnnoy_0.0.22            bslib_0.8.0                </span></span>
<span><span class="co">##  [95] irlba_2.3.5.1               KernSmooth_2.23-26         </span></span>
<span><span class="co">##  [97] rpart_4.1.24                colorspace_2.1-1           </span></span>
<span><span class="co">##  [99] DBI_1.2.3                   Hmisc_5.2-2                </span></span>
<span><span class="co">## [101] nnet_7.3-20                 tidyselect_1.2.1           </span></span>
<span><span class="co">## [103] bit_4.5.0.1                 compiler_4.4.2             </span></span>
<span><span class="co">## [105] curl_6.0.1                  httr2_1.0.7                </span></span>
<span><span class="co">## [107] htmlTable_2.4.3             xml2_1.3.6                 </span></span>
<span><span class="co">## [109] desc_1.4.3                  DelayedArray_0.32.0        </span></span>
<span><span class="co">## [111] plotly_4.10.4               checkmate_2.3.2            </span></span>
<span><span class="co">## [113] scales_1.3.0                lmtest_0.9-40              </span></span>
<span><span class="co">## [115] rappdirs_0.3.3              goftest_1.2-3              </span></span>
<span><span class="co">## [117] stringr_1.5.1               digest_0.6.37              </span></span>
<span><span class="co">## [119] minqa_1.2.8                 spatstat.utils_3.1-2       </span></span>
<span><span class="co">## [121] reader_1.0.6                rmarkdown_2.29             </span></span>
<span><span class="co">## [123] htmltools_0.5.8.1           pkgconfig_2.0.3            </span></span>
<span><span class="co">## [125] jpeg_0.1-10                 base64enc_0.1-3            </span></span>
<span><span class="co">## [127] lme4_1.1-36                 MatrixGenerics_1.18.1      </span></span>
<span><span class="co">## [129] dbplyr_2.5.0                fastmap_1.2.0              </span></span>
<span><span class="co">## [131] rlang_1.1.4                 htmlwidgets_1.6.4          </span></span>
<span><span class="co">## [133] UCSC.utils_1.2.0            shiny_1.10.0               </span></span>
<span><span class="co">## [135] farver_2.1.2                jquerylib_0.1.4            </span></span>
<span><span class="co">## [137] zoo_1.8-12                  jsonlite_1.8.9             </span></span>
<span><span class="co">## [139] BiocParallel_1.40.0         R.oo_1.27.0                </span></span>
<span><span class="co">## [141] VariantAnnotation_1.52.0    RCurl_1.98-1.16            </span></span>
<span><span class="co">## [143] magrittr_2.0.3              Formula_1.2-5              </span></span>
<span><span class="co">## [145] GenomeInfoDbData_1.2.13     dotCall64_1.2              </span></span>
<span><span class="co">## [147] patchwork_1.3.0             munsell_0.5.1              </span></span>
<span><span class="co">## [149] Rcpp_1.0.14                 reticulate_1.40.0          </span></span>
<span><span class="co">## [151] stringi_1.8.4               zlibbioc_1.52.0            </span></span>
<span><span class="co">## [153] MASS_7.3-64                 plyr_1.8.9                 </span></span>
<span><span class="co">## [155] parallel_4.4.2              listenv_0.9.1              </span></span>
<span><span class="co">## [157] ggrepel_0.9.6               deldir_2.0-4               </span></span>
<span><span class="co">## [159] splines_4.4.2               tensor_1.5                 </span></span>
<span><span class="co">## [161] hms_1.1.3                   igraph_2.0.3               </span></span>
<span><span class="co">## [163] spatstat.geom_3.3-4         RcppHNSW_0.6.0             </span></span>
<span><span class="co">## [165] reshape2_1.4.4              biomaRt_2.62.0             </span></span>
<span><span class="co">## [167] XML_3.99-0.17               evaluate_1.0.3             </span></span>
<span><span class="co">## [169] latticeExtra_0.6-30         biovizBase_1.54.0          </span></span>
<span><span class="co">## [171] renv_1.0.11                 BiocManager_1.30.25        </span></span>
<span><span class="co">## [173] NCmisc_1.2.0                nloptr_2.1.1               </span></span>
<span><span class="co">## [175] tweenr_2.0.3                httpuv_1.6.15              </span></span>
<span><span class="co">## [177] RANN_2.6.2                  tidyr_1.3.1                </span></span>
<span><span class="co">## [179] purrr_1.0.2                 polyclip_1.10-7            </span></span>
<span><span class="co">## [181] future_1.34.0               scattermore_1.2            </span></span>
<span><span class="co">## [183] ggforce_0.4.2               monocle3_1.3.7             </span></span>
<span><span class="co">## [185] xtable_1.8-4                restfulr_0.0.15            </span></span>
<span><span class="co">## [187] RSpectra_0.16-2             later_1.4.1                </span></span>
<span><span class="co">## [189] glasso_1.11                 viridisLite_0.4.2          </span></span>
<span><span class="co">## [191] ragg_1.3.3                  tibble_3.2.1               </span></span>
<span><span class="co">## [193] memoise_2.0.1               GenomicAlignments_1.42.0   </span></span>
<span><span class="co">## [195] cluster_2.1.8               globals_0.16.3</span></span></code></pre>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xuan Liang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
